### DISCUSSION TITLE
collectionSpec.Variables miss static volatile const var

### DISCUSSION THREAD

**Discussion by Ghostbaby** [cilium/ebpf#1620] [Q&A] [answered]
- go.mod: `github.com/cilium/ebpf v0.16.1-0.20241204125435-9895aae6467e`
- ebpf code:
```c
static volatile const struct config CFG;
#define cfg (&CFG)
```

- go code:
```go
	// 将配置写入到 bpf 程序中
	if err := bpfSpecClone.RewriteConstants(map[string]interface{}{
		"CFG": traceConfig,
	}); err != nil {
		log.Fatalf("Failed to rewrite config: %v", err)
	}
```
- error:
```bash
F1205 17:36:24.336716  215397 log.go:130] Failed to rewrite config: rewrite constants: some constants are missing from .rodata: CFG
```
- build code:
```make
cd ./cmd;CGO_ENABLED=1 GOOS=linux GOARCH=amd64 CGO_LDFLAGS='-g -lcapstone -static'   go build -gcflags "all=-N -l" -v  -o nfs-trace
```

**Comment by ti-mo** [comment, marked as answer]
I understand this may be a breaking change, but this was unavoidable as RewriteConstants didn't take object visibility into account in the past (it allowed rewriting local vars). With VariableSpec in fhe picture, we now follow the same rules as libbpf.

Please see https://ebpf-go.dev/concepts/global-variables/#static-global-variables for more info.

Also, in the future, please create issues for reporting regressions, Discussions are more for Q&A.
